### PR TITLE
[GraphBolt] Support numpy for edges when constructing graph in `preprocess_ondisk_dataset`.

### PR DESCRIPTION
**Pull request by yxy235**
## Description

Support numpy for edges when constructing graph in `preprocess_ondisk_dataset`.
**Using numpy for edges in YAML file:**
```
edges:
    - format: numpy
      path: your_edge_path.npy
```

### Benchmark
TBD.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 1]
```diff
@@ -86,13 +86,22 @@ def preprocess_ondisk_dataset(
...
         # Homogeneous graph.
         num_nodes = input_config["graph"]["nodes"][0]["num"]
-        edge_data = pd.read_csv(
-            os.path.join(
-                dataset_dir, input_config["graph"]["edges"][0]["path"]
-            ),
-            names=["src", "dst"],
-        )
-        src, dst = edge_data["src"].to_numpy(), edge_data["dst"].to_numpy()
+        if input_config["graph"]["edges"][0]["format"] == "numpy":
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:89]: add sanity check here that only numpy and csv are supported.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 2]
```diff
@@ -86,13 +86,22 @@ def preprocess_ondisk_dataset(
...
+            edge_data = read_data(
+                os.path.join(
+                    dataset_dir, input_config["graph"]["edges"][0]["path"]
+                ),
+                input_config["graph"]["edges"][0]["format"],
+            )
+            src, dst = edge_data
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:96]: assert shape here that requires to be `(2, N)`.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 3]
```diff
@@ -104,12 +113,19 @@ def preprocess_ondisk_dataset(
         # Construct the data dict.
         data_dict = {}
         for edge_info in input_config["graph"]["edges"]:
-            edge_data = pd.read_csv(
-                os.path.join(dataset_dir, edge_info["path"]),
-                names=["src", "dst"],
-            )
-            src = torch.tensor(edge_data["src"])
-            dst = torch.tensor(edge_data["dst"])
+            if edge_info["format"] == "numpy":
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:116]: is it possible to remove the code duplication here which is same as in homograph code path?

File: tests/python/pytorch/graphbolt/gb_test_utils.py [hunk 4]
```diff
@@ -88,22 +88,28 @@ def random_hetero_graph(num_nodes, num_edges, num_ntypes, num_etypes):
...
+    if edge_fmt == "csv":
+        # Wrtie into edges/edge.csv
+        edges = pd.DataFrame(edges, columns=["src", "dst"])
+        edge_path = os.path.join("edges", "edge.csv")
+        edges.to_csv(
+            os.path.join(test_dir, edge_path),
+            index=False,
+            header=False,
+        )
+    else:
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/gb_test_utils.py:108]: `else if edge_fmt == "numpy"` or assert in advance

**Review by Rhett-Ying — COMMENTED** [review]
pls update the tutorials accordingly in https://docs.dgl.ai/en/latest/stochastic_training/ondisk-dataset.html. let's do it in another PR.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 5]
```diff
@@ -86,14 +90,12 @@ def preprocess_ondisk_dataset(
     if is_homogeneous:
         # Homogeneous graph.
         num_nodes = input_config["graph"]["nodes"][0]["num"]
-        edge_data = pd.read_csv(
-            os.path.join(
-                dataset_dir, input_config["graph"]["edges"][0]["path"]
-            ),
-            names=["src", "dst"],
+        edge_fmt = input_config["graph"]["edges"][0]["format"]
+        assert edge_fmt in ["numpy", "csv"], print(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:94]: moving such sanity check into `read_edges` is more reasonable and so that code here is simpler.

Comment (Rhett-Ying) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:94]: if more formats are added in the future, we just need to update `read_edges()`.

Comment (Skeleton003) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:94]: why not do the assertion in `read_edges()`?

File: python/dgl/graphbolt/internal/utils.py [hunk 6]
```diff
@@ -120,3 +121,25 @@ def get_attributes(_obj) -> list:
         and not callable(getattr(_obj, attribute))
     ]
     return attributes
+
+
+def read_edges(dataset_dir, edge_fmt, edge_path):
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/internal/utils.py:126]: add dedicated testcase for unsupported format and capture exception with matched log.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 5 above)
Comment (Skeleton003) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:94]: BTW, no need to `print()`, just
```python
assert edge_fmt in ["numpy", "csv"], "only numpy and csv are supported for edges."
```
should be ok

File: python/dgl/graphbolt/internal/utils.py [hunk 7]
```diff
@@ -120,3 +121,28 @@ def get_attributes(_obj) -> list:
...
+    """Read egde data from numpy or csv."""
+    assert edge_fmt in ["numpy", "csv"], print(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/internal/utils.py:128]: `print()` is not needed.

File: python/dgl/graphbolt/internal/utils.py [hunk 8]
```diff
@@ -120,3 +121,28 @@ def get_attributes(_obj) -> list:
...
+        "Only numpy and csv are supported for edges."
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/internal/utils.py:129]:
```suggestion
        f"`numpy` or `csv` is expected when reading edges but got {edge_fmt}."
```

File: python/dgl/graphbolt/internal/utils.py [hunk 9]
```diff
@@ -120,3 +121,28 @@ def get_attributes(_obj) -> list:
...
+    )
+    if edge_fmt == "numpy":
+        edge_data = read_data(
+            os.path.join(dataset_dir, edge_path),
+            edge_fmt,
+        )
+        assert edge_data.shape[0] == 2 and len(edge_data.shape) == 2, print(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/internal/utils.py:136]: ditto: `print()`. And print out what shape is passed in

File: tests/python/pytorch/graphbolt/utils/test_internal.py [hunk 10]
```diff
@@ -141,3 +143,58 @@ def test_copy_or_convert_data(data_fmt, save_fmt, is_feature):
...
+            edges = edges.T
+            edge_path = os.path.join("edges", "edge.npy")
+            np.save(os.path.join(test_dir, edge_path), edges)
+        src, dst = internal.read_edges(test_dir, edge_fmt, edge_path)
+        assert torch.equal(src, torch.tensor(nodes))
+        assert torch.equal(dst, torch.tensor(neighbors))
+
+
+def test_read_edges_error():
+    with pytest.raises(
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/utils/test_internal.py:177]: add comment: `1. Unsupported file format.`

File: tests/python/pytorch/graphbolt/utils/test_internal.py [hunk 11]
```diff
@@ -141,3 +143,58 @@ def test_copy_or_convert_data(data_fmt, save_fmt, is_feature):
...
+        AssertionError,
+        match="`numpy` or `csv` is expected when reading edges but got `fake-type`.",
+    ):
+        internal.read_edges("test_dir", "fake-type", "edge_path")
+
+    with tempfile.TemporaryDirectory() as test_dir:
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/utils/test_internal.py:183]: `2. Unexpected shape of numpy array`

**Review by Rhett-Ying — APPROVED** [review]

**Comment by yxy235** [comment]
@dgl-bot

**Pull request merged**